README.md

```diff
@@ -19,6 +20,59 @@ model = torch.nn.Sequential(a, b, c, d)
 model = fairscale.nn.Pipe(model, balance=[2, 2], devices=[0, 1], chunks=8)
 ```
 
+### Optimizer state sharding (ZeRO)
+See a more complete example [here](https://github.com/facebookresearch/fairscale/blob/oss_async_broadcast/benchmarks/oss.py), but a minimal example could look like the following :
+
+```bash
+import torch
+from fairscale.optim.oss import OSS
+
+def train(
+    rank: int,
+    world_size: int,
+    epochs: int):
+
+    # DDP
+    dist_init(rank, world_size)
+
+    # Problem statement
+    model = myAwesomeModel()
+    dataloader = mySuperFastDataloader()
+    loss = myVeryRelevantLoss()
+    base_optimizer = torch.optim.SGD # pick any pytorch compliant optimizer here
+    base_optimizer_arguments = {} # pass any optimizer specific arguments here, or directly below when instantiating OSS
+
+    optimizer = OSS(params=model.parameters(), optim=base_optimizer, **base_optimizer_arguments)
+
+    # Any relevant training loop, nothing specific to OSS. For example:
+    model.train()
+    for e in range(epochs):
+        for batch in dataloader:
+            def closure():
+                model.zero_grad()
+                outputs = model(batch["inputs"])
+                loss = loss_fn(outputs, batch["label"])
+                torch.distributed.all_reduce(loss, op=torch.distributed.ReduceOp.SUM)
+                loss /= world_size
+                loss.backward()
+                return loss
+
+            optimizer.step(closure)
```

Comment:
using closure is not the most common way for pytorch optimizers, right? Perhaps this example can be simplified by not using closure?

Reply:
depends on people, but closures is (to my understanding) considered a little safer because the scope is tight. The scope in python is leaky (like, something defined in a for loop leaks outside of it), so closures bring some sanity. It also makes it compatible with optimizers which require multiple evaluations, both options (with and without closures) are in the pytorch doc https://pytorch.org/docs/stable/optim.html#taking-an-optimization-step

Reply:
Completely agree. I was just suggesting that it is not the most common way people do it (AFAIK) (perhaps because most programmers are not very used to it). For an initial example, perhaps keep it simple is better. Of course, if you want to use this opportunity to advocate the usage of closure style, it is fine too. In that case, perhaps some comments in the code to explain what's going on with closure and why it is better?

Reply:
follow up: closure removed !

Reply:
I think it's an orthogonal subject to zero, so probably best to not mix things up. Fine by me as it is, no worries, I'll keep using closures for my own code though :)